File: 46-permutations/46-permutations.cpp
class Solution {
public:
    void generate(vector<vector<int>>& output, vector<int>& v, unordered_map<int,int>& m, vector<int>& nums){
        if(v.size() == nums.size()){
            output.push_back(v);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            if(!m[i]){
                m[i] = 1;
                v.push_back(nums[i]);
                generate(output,v,m,nums);
                v.pop_back();
                m[i] = 0 ;
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> output;
        vector<int> v;
        unordered_map<int,int> m;
        generate(output,v,m,nums);
        return output;
    }
};

File: 46-permutations/46-permutations.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public void dfs(int[] nums, ArrayList<Integer> arr, boolean[] visited){
        if(arr.size() == nums.length){
            ans.add(new ArrayList<>(arr));
            return;
        }
        for(int i = 0; i < nums.length; i++){
            if(!visited[i]){
                visited[i] = true;
                arr.add(nums[i]);
                dfs(nums,arr,visited);
                arr.remove(arr.size()-1);
                visited[i] = false;
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        int len = nums.length;
        boolean[] visited = new boolean[len];
        ArrayList<Integer> arr = new ArrayList<>();
        dfs(nums,arr,visited);
        return ans;
    }
}